src/encodings.jl

```diff
@@ -1,271 +1,384 @@
+# Parametric singleton type representing a given string encoding via its symbol parameter
+
+import Base: show, print, convert
+export Encoding, @enc_str, codeunit, native_endian
+
+immutable Encoding{enc} end
+
+Encoding(s) = Encoding{symbol(s)}()
+macro enc_str(s)
+    :(Encoding{$(Expr(:quote, symbol(s)))}())
+end
+
+convert{T<:AbstractString, enc}(::Type{T}, ::Encoding{enc}) = string(enc)
+
+show{enc}(io::IO, ::Encoding{enc}) = print(io, string(enc), " string encoding type")
+print{enc}(io::IO, ::Encoding{enc}) = print(io, enc)
+
+
+## Get the encoding used by a string type
+encoding(::Type{ASCIIString}) = enc"ASCII"
+encoding(::Type{UTF8String})  = enc"UTF-8"
+
+if ENDIAN_BOM == 0x04030201
+    encoding(::Type{UTF16String}) = enc"UTF-16LE"
+    encoding(::Type{UTF32String}) = enc"UTF-32LE"
+else
+    encoding(::Type{UTF16String}) = enc"UTF-16BE"
+    encoding(::Type{UTF32String}) = enc"UTF-32BE"
+end
+
+
+## Functions giving information about a particular encoding
+
+"""
+    native_endian(enc)
+
+    Check whether string encoding `enc` follows the current machine endianness.
+    `enc` can be specified either as a string or as an `Encoding` object.
+"""
+native_endian(::Encoding) = true
+
+if ENDIAN_BOM == 0x04030201
+    native_endian(::Encoding{symbol("UTF-16LE")}) = true
+    native_endian(::Encoding{symbol("UTF-32LE")}) = true
+    native_endian(::Encoding{symbol("UTF-16BE")}) = false
+    native_endian(::Encoding{symbol("UTF-32BE")}) = false
+
+    native_endian(::Encoding{symbol("UTF16LE")}) = true
+    native_endian(::Encoding{symbol("UTF32LE")}) = true
+    native_endian(::Encoding{symbol("UTF16BE")}) = false
+    native_endian(::Encoding{symbol("UTF32BE")}) = false
+else
+    native_endian(::Encoding{symbol("UTF-16LE")}) = false
+    native_endian(::Encoding{symbol("UTF-32LE")}) = false
+    native_endian(::Encoding{symbol("UTF-16BE")}) = true
+    native_endian(::Encoding{symbol("UTF-32BE")}) = true
+
+    native_endian(::Encoding{symbol("UTF16LE")}) = false
+    native_endian(::Encoding{symbol("UTF32LE")}) = false
+    native_endian(::Encoding{symbol("UTF16BE")}) = true
+    native_endian(::Encoding{symbol("UTF32BE")}) = true
+end
+
+native_endian(enc::AbstractString) = native_endian(Encoding(enc))
+
+
+"""
+    codeunit(enc)
+
+    Get the type corresponding to a code unit in the encoding `enc`.
+    `enc` can be specified either as a string or as an `Encoding` object.
+"""
+@generated function codeunit{enc_s}(enc::Encoding{enc_s})
+    s = string(enc_s)
+    if s in encodings32
+       :UInt32
+    elseif s in encodings16
+       :UInt16
+    elseif s in encodings8
+       :UInt8
+    else
+       error("encoding with unknown codeunit size, define a codeunit method for it")
+    end
+end
+
+codeunit(enc::AbstractString) = codeunit(Encoding(enc))
+
+
+## Lists of all known encodings taken from various iconv implementations,
+## including different aliases for the same encoding
+
+# 8-bit codeunit encodings
+const encodings8 = [
```

Comment:
Could that be done at compile time for most cases?  Esp. with a string macro for indicating the encodings?  For places that get an encoding name passed in, it is used multiple times and performance matters, people should simply be told to create an encoding object and then call a function that can be specialized on that encoding to do the work.